<?xml version="1.0" encoding="UTF-8"?><!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml" lang="pt"><head><meta http-equiv="Content-Type" content="text/html;charset=UTF-8"/><link rel="stylesheet" href="../jacoco-resources/report.css" type="text/css"/><link rel="shortcut icon" href="../jacoco-resources/report.gif" type="image/gif"/><title>Genoma.java</title><link rel="stylesheet" href="../jacoco-resources/prettify.css" type="text/css"/><script type="text/javascript" src="../jacoco-resources/prettify.js"></script></head><body onload="window['PR_TAB_WIDTH']=4;prettyPrint()"><div class="breadcrumb" id="breadcrumb"><span class="info"><a href="../jacoco-sessions.html" class="el_session">Sessions</a></span><a href="../index.html" class="el_report">GeneDecoder</a> &gt; <a href="index.source.html" class="el_package">com.study.oop</a> &gt; <span class="el_source">Genoma.java</span></div><h1>Genoma.java</h1><pre class="source lang-java linenums">package com.study.oop;

import java.io.File;
import java.io.FileNotFoundException;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Scanner;
import java.util.TreeMap;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Classe que implementa um {@link Genoma}.
 *
 * @author &lt;a href=&quot;mailto:yuri.arend@acad.pucrs.br&quot;&gt;yuri.arend&lt;/a&gt;
 * @since 15/09/2018 14:37:00
 */
public class Genoma {
    private Map&lt;String, Gene&gt; geneMap;

<span class="fc" id="L23">    public Genoma(String filePath) {</span>
<span class="fc" id="L24">        geneMap = new TreeMap&lt;&gt;();</span>
        try {
<span class="fc" id="L26">            loadDataFrom(new File(filePath));</span>
<span class="nc" id="L27">        } catch (FileNotFoundException e) {</span>
<span class="nc" id="L28">            e.printStackTrace();</span>
<span class="fc" id="L29">        }</span>
<span class="fc" id="L30">    }</span>

    public Map&lt;String, Gene&gt; getGeneMap() {
<span class="fc" id="L33">        return geneMap;</span>
    }

    /**
     * Método que efetua a leitura de um arquivo FASTA e faz a montagem do geneMap
     * conforme faz a leitura dos genes no arquivo.
     *
     * @param filePath caminho do arquivo FASTA no disco
     * @throws FileNotFoundException exception caso não encontre o arquivo no caminho especificado
     */
    private void loadDataFrom(File filePath) throws FileNotFoundException {
<span class="fc" id="L44">        Pattern pattern = Pattern.compile(&quot;(\\[(.*?)\\])&quot;);</span>
<span class="fc" id="L45">        Pattern patternLocation = Pattern.compile(&quot;(\\[location=(\\d+)..(\\d+)\\])&quot;);</span>
<span class="fc" id="L46">        Pattern patternLocusTag = Pattern.compile(&quot;(\\[locus_tag=(.*?)\\])&quot;);</span>

<span class="fc" id="L48">        try (Scanner data = new Scanner(filePath)) {</span>
<span class="fc" id="L49">            boolean first = true;</span>
<span class="fc" id="L50">            String locusTag = null;</span>
<span class="fc" id="L51">            long begin = -1;</span>
<span class="fc" id="L52">            long end = -1;</span>
<span class="fc" id="L53">            List&lt;Character&gt; dnaBasesSequence = new ArrayList&lt;&gt;();</span>
            String line;

<span class="fc bfc" id="L56" title="All 2 branches covered.">            while (data.hasNextLine()) {</span>
<span class="fc" id="L57">                line = data.nextLine();</span>
<span class="fc bfc" id="L58" title="All 2 branches covered.">                if (isHeaderLine(line)) {</span>
<span class="fc bfc" id="L59" title="All 2 branches covered.">                    if (first) {</span>
<span class="fc" id="L60">                        first = false;</span>
                    } else {
<span class="fc" id="L62">                        addGeneInGenoma(locusTag, begin, end, dnaBasesSequence);</span>
<span class="fc" id="L63">                        locusTag = &quot;&quot;;</span>
<span class="fc" id="L64">                        begin = -1;</span>
<span class="fc" id="L65">                        end = -1;</span>
<span class="fc" id="L66">                        dnaBasesSequence.clear();</span>
                    }
<span class="fc" id="L68">                    Matcher patternMatcher = pattern.matcher(line);</span>
<span class="fc bfc" id="L69" title="All 2 branches covered.">                    while (patternMatcher.find()) {</span>
<span class="fc" id="L70">                        String token = patternMatcher.group(1);</span>
<span class="fc" id="L71">                        Matcher patterLocationMatcher = patternLocation.matcher(token);</span>
<span class="fc bfc" id="L72" title="All 2 branches covered.">                        if (patterLocationMatcher.matches()) {</span>
<span class="fc" id="L73">                            begin = Long.parseLong(patterLocationMatcher.group(2));</span>
<span class="fc" id="L74">                            end = Long.parseLong(patterLocationMatcher.group(3));</span>
                        } else {
<span class="fc" id="L76">                            Matcher patternLocusTagMatcher = patternLocusTag.matcher(token);</span>
<span class="fc bfc" id="L77" title="All 2 branches covered.">                            if (patternLocusTagMatcher.matches()) {</span>
<span class="fc" id="L78">                                locusTag = patternLocusTagMatcher.group(2);</span>
                            }
                        }
<span class="fc" id="L81">                    }</span>
<span class="fc" id="L82">                } else {</span>
<span class="fc" id="L83">                    mountSequence(dnaBasesSequence, line);</span>
                }

            }
<span class="fc" id="L87">            addGeneInGenoma(locusTag, begin, end, dnaBasesSequence);</span>
        }


<span class="fc" id="L91">    }</span>

    /**
     * Método que adiciona o com.study.oop.Gene lido no arquivo FASTA para dentro do com.study.oop.Genoma.
     *
     * @param locus    locus tag do gene extraída do arquivo FASTA
     * @param begin    início de DNA do gene extraída do arquivo FASTA
     * @param end      final do DNA do gene extraída do arquivo FASTA
     * @param sequence sequência de DNA do gene extraída do arquivo FASTA
     */
    private void addGeneInGenoma(String locus, long begin, long end, List&lt;Character&gt; sequence) {
<span class="fc" id="L102">        Gene gene = new Gene(locus, begin, end, sequence);</span>
<span class="fc" id="L103">        geneMap.put(locus, gene);</span>
<span class="fc" id="L104">    }</span>

    /**
     * Método que monta a sequência de DNA do gene caractere à caractere da linha.
     *
     * @param sequence lista onde serão adicionados os DNAS
     * @param line     linha atual do arquivo FASTA
     */
    private void mountSequence(List&lt;Character&gt; sequence, String line) {
<span class="fc bfc" id="L113" title="All 2 branches covered.">        for (int i = 0; i &lt; line.length(); i++) {</span>
<span class="fc" id="L114">            sequence.add(line.charAt(i));</span>
        }
<span class="fc" id="L116">    }</span>

    /**
     * Método para verificar se a linha atual do arquivo FASTA corresponde a
     * uma linha de cabeçalho.
     *
     * @param line linha a ser verificada
     * @return true caso seja uma linha de cabeçalho de gene.
     */
    private boolean isHeaderLine(String line) {
<span class="fc bfc" id="L126" title="All 4 branches covered.">        return line.length() &gt; 0 &amp;&amp; Objects.equals('&gt;', line.charAt(0));</span>
    }
}
</pre><div class="footer"><span class="right">Created with <a href="http://www.jacoco.org/jacoco">JaCoCo</a> 0.8.2.201808211720</span></div></body></html>